src/audio/mfcc/mfcc_generic.c

```diff
@@ -0,0 +1,444 @@
+	m = buf->s_avail / fft->fft_hop_size;
+	for (i = 0; i < m; i++) {
+		/* Clear FFT input buffer because it has been used as scratch */
+		bzero(fft->fft_buf, fft->fft_buffer_size);
+
+		/* Copy data to FFT input buffer from overlap buffer and from new samples buffer */
+		mfcc_fill_fft_buffer(state);
+
+		/* TODO: remove_dc_offset */
+
+		/* TODO: use_energy & raw_energy */
+
+#ifdef MFCC_NORMALIZE_FFT
+		/* Find block scale left shift for FFT input */
+		input_shift = mfcc_normalize_fft_buffer(state);
+#else
+		input_shift = 0;
+#endif
+
+		/* Window function */
+		mfcc_apply_window(state, input_shift);
+
+		/* TODO: use_energy & !raw_energy */
+
+#ifdef DEBUGFILES
+		for (j = 0; j < fft->fft_padded_size; j++)
+			fprintf(fh_fft_in, "%d %d\n", fft->fft_buf[j].real, fft->fft_buf[j].imag);
+#endif
+
+		/* The FFT out buffer needs to be cleared to avoid to corrupt
+		 * the output. TODO: check moving it to FFT lib.
+		 */
+		bzero(fft->fft_out, fft->fft_buffer_size);
+
+		/* Compute FFT */
+#if MFCC_FFT_BITS == 16
+		fft_execute_16(fft->fft_plan, false);
+#else
+		fft_execute_32(fft->fft_plan, false);
+#endif
+
+#ifdef DEBUGFILES
+		for (j = 0; j < fft->fft_padded_size; j++)
+			fprintf(fh_fft_out, "%d %d\n", fft->fft_out[j].real, fft->fft_out[j].imag);
+#endif
+
+		/* Convert powerspectrum to Mel band logarithmic spectrum */
+		mat_init_16b(state->mel_spectra, 1, state->dct.num_mel, 7); /* Q8.7 */
+#if MFCC_FFT_BITS == 16
+		psy_apply_mel_filterbank_16(&state->melfb, fft->fft_out, state->power_spectra,
+					    state->mel_spectra->data, input_shift);
+#else
+		psy_apply_mel_filterbank_32(&state->melfb, fft->fft_out, state->power_spectra,
+					    state->mel_spectra->data, input_shift);
+#endif
+
+		/* Multiply Mel spectra with DCT matrix to get cepstral coefficients */
+		mat_init_16b(state->cepstral_coef, 1, state->dct.num_ceps, 7); /* Q8.7 */
+		mat_multiply(state->mel_spectra, state->dct.matrix, state->cepstral_coef);
+
+		/* Apply cepstral lifter */
+		if (state->lifter.cepstral_lifter != 0)
+			mat_multiply_elementwise(state->cepstral_coef, state->lifter.matrix,
+						 state->cepstral_coef);
+
+		cc_count += state->dct.num_ceps;
+
+		/* Output to sink buffer */
+
+#ifdef DEBUGFILES
+		for (j = 0; j < fft->fft_padded_size; j++)
+			fprintf(fh_pow, "%d\n", state->power_spectra[j]);
+
+		for (j = 0; j < state->dct.num_mel; j++)
+			fprintf(fh_mel, " %d", state->mel_spectra->data[j]);
+
+		fprintf(fh_mel, "\n");
+		for (j = 0; j < state->dct.num_ceps; j++)
+			fprintf(fh_ceps, " %d", state->cepstral_coef->data[j]);
+
+		fprintf(fh_ceps, "\n");
+#endif
+	}
+
+	/* TODO: This version handles only one FFT run per copy(). How to pass multiple
+	 * cepstral coefficients sets return is an open.
+	 */
+	return cc_count;
+}
+
+#if CONFIG_FORMAT_S16LE
+
+static int16_t *mfcc_sink_copy_zero_s16(const struct audio_stream *sink,
+					int16_t *w_ptr, int samples)
+{
+	int nmax;
+	int i;
+	int n;
+	int copied = 0;
+
+	while (copied < samples) {
```

Comment:
in many locations this can be `for (copied = 0; copied < samples; copied += n)`

Reply:
Done now, thanks!